docs/python-api/triton-semantics.rst

```diff
@@ -0,0 +1,42 @@
+
+The algorithm is as follows:
+
+1. **Kind** If one tensor is of a dtype of a higher kind, the other tensor is promoted to this dtype: ``(int32, bfloat16) -> bfloat16``
+
+2. **Width** If both tensors are of dtypes of the same kind, and one of them is of a higher width, the other one is promoted to this dtype: ``(float32, float16) -> float32``
+
+3. **Supremum** If both tensors are of the same width and signedness but different dtypes, they are both promoted to the next larger dtype. ``(float16, bfloat16) -> float32``
+
+   3.1 If both tensors are of different ``fp8`` dtypes, they are both cast to ``float16``.
+
+4. **Prefer unsigned** Otherwise (same width, different signedness), they are promoted to the unsigned dtype: ``(int32, uint32) -> uint32``
+
+When an operation involves a tensor and a scalar
+
+1. If the scalar is of a kind lower or equal to the tensor, it will not participate in the promotion: ``(uint8, int) -> uint8``
+
+2. If the scalar is of a higher kind, we choose the lowest dtype in which it fits among ``int32`` < ``uint32`` < ``int64`` < ``uint64`` for ints and ``float32`` < ``float64`` for floats. Then, both the tensor and the scalar are promoted to this dtype: ``(int16, 4.0) -> float32``
+
+
+Broadcasting
+============
+
+**Broadcasting** allows operations on tensors of different shapes by automatically expanding their shapes to a compatible size without copying the data. This follows the following rules:
+
+1. If one of the tensor shapes is shorter, pad it on the left with ones until both tensors have the same number of dimensions: ``((3, 4), (5, 3, 4)) -> ((1, 3, 4), (5, 3, 4))``
+
+2. Two dimensions are compatible if they are equal, or if one of them is 1. A dimension of 1 will be expanded to match the dimension of the other tensor. ``((1, 3, 4), (5, 3, 4)) -> ((5, 3, 4), (5, 3, 4))``
+
+
+Differences with NumPy
+======================
+
+**C rounding in integer division** Operators in Triton follow C semantics rather than Python semantics for efficiency. As such, ``int // int`` implements `rounding towards zero as in C <https://en.wikipedia.org/wiki/Modulo#In_programming_languages>`_ for integers of mixed signs, rather than rounding towards minus infinity as in Python. For the same reason, the modulus operator ``int % int`` (which is defined as ``a % b = a - b * (a // b)``) also follows C semantics rather than Python semantics.
+
+Perhaps confusingly, integer division and modulus follow Python semantics for computations stored in variables decorated with the ``tl.constexpr`` type.
```

Comment:
> stored in variables decorated with the ``tl.constexpr`` type.

thinking again, perhaps this should be "scalar" as you use `scalar == constexpr` throughout the document.  Or perhaps it should be "scalar constant" throughout.

Reply:
not quite. `tl.constexpr` is only propagated if you tag a given variable as `tl.constexpr`. As such, the semantics are rather confusing. For example, if you have `a` constexpr and write `b = a % 32`, `b` is not constexpr, it is cast to a 0-D tensor. For it to be constexpr you have to mark it as `b: tl.constexpr = a % 32`. With this in mind, it's trick to balance this paragraph and the one about scalars without describing what are the rules for the propagation of `tl.constexpr`.


I tried defining scalar a few paragraphs before, and discussed this in terms of operations on scalar inputs. Tell me what you think

I'm not sure if the behaviour above is expected (I would assume it is, as there is specific code that makes this behaviour happen), but I find it particularly confusing. cc @ptillet

Reply:
It looks like the assignment operation automatically casts to tensor.
https://github.com/triton-lang/triton/blob/0e3cadda928f3946e1fedaff8608ddb0b093a941/python/triton/compiler/code_generator.py#L501-L507


IMO that's pretty surprising as IIUC
```
INT_MAX * 2
```
gives a uint32 whereas
```
a = INT_MAX
b = a * 2
```
overflows an int32.

Reply:
Or another example `-1 // 2 == -1` but `a = -1; a // 2 == 0`

Reply:
thoughts @ThomasRaoux @ptillet?